src/gurt/misc.c

```diff
@@ -13,26 +13,482 @@
+
+/** TLS memory allocation counter */
+struct dm_tls_counter {
+	/** already allocated size */
+	int64_t		 mtc_size;
+	/** total number of calls of allocation */
+	int64_t		 mtc_count;
+	/** true if this TLS counter is registered on \a dm_counters */
+	bool		 mtc_registered;
+};
+
+/** thread-local counters */
+static __thread struct dm_tls_counter dm_tls_counters[M_TAG_MAX];
+static pthread_mutex_t	dm_lock = PTHREAD_MUTEX_INITIALIZER;
+
+/** Assuming no more than 128 xstreams */
+#define DM_TLS_MAX	128
+
+/** All the defined counters */
+struct dm_counters {
+	/** counter tag, see \a dm_tag */
+	int			 mc_tag;
+	/** index of the last TLS counter in \a mc_tls_cntrs */
+	int			 mc_last;
+	/** counter string name */
+	char			*mc_name;
+	/** all the registered tls counters */
+	struct dm_tls_counter	*mc_tls_cntrs[DM_TLS_MAX];
+};
+
+static struct dm_counters	dm_counters[] = {
+	{
+		.mc_tag		= M_OTHER,
+		.mc_name	= "other",
+	},
+	{
+		.mc_tag		= M_STR,
+		.mc_name	= "string",
+	},
+	{
+		.mc_tag		= M_RANK,
+		.mc_name	= "rank",
+	},
+	{
+		.mc_tag		= M_CRT,
+		.mc_name	= "cart",
+	},
+	{
+		.mc_tag		= M_CRT_RPC,
+		.mc_name	= "cart_rpc",
+	},
+	{
+		.mc_tag		= M_CRT_IV,
+		.mc_name	= "cart_iv",
+	},
+	{
+		.mc_tag		= M_SWIM,
+		.mc_name	= "swim",
+	},
+	{
+		.mc_tag		= M_CSUM,
+		.mc_name	= "csum",
+	},
+	{
+		.mc_tag		= M_PROP,
+		.mc_name	= "prop",
+	},
+	{
+		.mc_tag		= M_TSE,
+		.mc_name	= "task",
+	},
+	{
+		.mc_tag		= M_PL,
+		.mc_name	= "pl",
+	},
+	{
+		.mc_tag		= M_IV,
+		.mc_name	= "incast",
+	},
+	{
+		.mc_tag		= M_SCHED,
+		.mc_name	= "sched",
+	},
+	{
+		.mc_tag		= M_IO,
+		.mc_name	= "io",
+	},
+	{
+		.mc_tag		= M_IO_ARG,
+		.mc_name	= "io_arg",
+	},
+	{
+		.mc_tag		= M_BIO,
+		.mc_name	= "bio",
+	},
+	{
+		.mc_tag		= M_VEA,
+		.mc_name	= "vea",
+	},
+	{
+		.mc_tag		= M_AGG,
+		.mc_name	= "agg",
+	},
+	{
+		.mc_tag		= M_RECOV,
+		.mc_name	= "recov",
+	},
+	{
+		.mc_tag		= M_EC,
+		.mc_name	= "ec",
+	},
+	{
+		.mc_tag		= M_EC_AGG,
+		.mc_name	= "ec_agg",
+	},
+	{
+		.mc_tag		= M_EC_RECOV,
+		.mc_name	= "ec_recov",
+	},
+	{
+		.mc_tag		= M_POOL,
+		.mc_name	= "pool",
+	},
+	{
+		.mc_tag		= M_CONT,
+		.mc_name	= "cont",
+	},
+	{
+		.mc_tag		= M_OBJ,
+		.mc_name	= "obj",
+	},
+	{
+		.mc_tag		= M_DTX,
+		.mc_name	= "dtx",
+	},
+	{
+		.mc_tag		= M_VOS,
+		.mc_name	= "vos",
+	},
+	{
+		.mc_tag		= M_VOS_TS,
+		.mc_name	= "vos_ts",
+	},
+	{
+		.mc_tag		= M_VOS_LRU,
+		.mc_name	= "vos_lru",
+	},
+	{
+		.mc_tag		= M_VOS_DTX,
+		.mc_name	= "vos_dtx",
+	},
+	{
+		.mc_tag		= M_RDB,
+		.mc_name	= "rdb",
+	},
+	{
+		.mc_tag		= M_TAG_MAX,
+		.mc_name	= NULL,
+	},
+};
+
+/* register thread-local counter to the counter table, so we can
+ * collect the summary, see dm_tag_query().
+ */
+static void
+dm_tls_register(int tag, struct dm_tls_counter *mtc)
+{
+	struct dm_counters	*mc;
+
+	mc = &dm_counters[tag];
+	D_ASSERTF(tag == mc->mc_tag, "Mimatched tag %s: %d/%d\n",
+		  mc->mc_name, mc->mc_tag, tag);
+
+	pthread_mutex_lock(&dm_lock);
```

Comment:
the one question I have here is whether there is something in telemetry that fits this kind of counter

Reply:
I think we should add telemetry for these counters, probably in a separate patch